modules/generator/registry/gauge.go

```diff
@@ -0,0 +1,188 @@
+package registry
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"go.uber.org/atomic"
+)
+
+// this is mostly copied from counter
+
+type gauge struct {
+	metricName string
+
+	// seriesMtx is used to sync modifications to the map, not to the data in series
+	seriesMtx sync.RWMutex
+	series    map[uint64]*gaugeSeries
+
+	onAddSeries    func(count uint32) bool
+	onRemoveSeries func(count uint32)
+}
+
+type gaugeSeries struct {
+	// labelValueCombo should not be modified after creation
+	labels          []string
+	labelValueCombo []string
+	value           *atomic.Float64
+	lastUpdated     *atomic.Int64
+	// firstSeries is used to track if this series is new to the gauge.  This
+	// is used to ensure that new gauges being with 0, and then are incremented
+	// to the desired value.  This avoids Prometheus throwing away the first
+	// value in the series, due to the transition from null -> x.
+	firstSeries *atomic.Bool
+}
+
+var _ Gauge = (*gauge)(nil)
+var _ metric = (*gauge)(nil)
+
+const add = "add"
+const set = "set"
+
+func (gs *gaugeSeries) isNew() bool {
+	return gs.firstSeries.Load()
+}
+
+func (gs *gaugeSeries) registerSeenSeries() {
+	gs.firstSeries.Store(false)
+}
+
+func newGauge(name string, onAddSeries func(uint32) bool, onRemoveSeries func(count uint32)) *gauge {
+	if onAddSeries == nil {
+		onAddSeries = func(uint32) bool {
+			return true
+		}
+	}
+	if onRemoveSeries == nil {
+		onRemoveSeries = func(uint32) {}
+	}
+
+	return &gauge{
+		metricName:     name,
+		series:         make(map[uint64]*gaugeSeries),
+		onAddSeries:    onAddSeries,
+		onRemoveSeries: onRemoveSeries,
+	}
+}
+
+func (g *gauge) Set(labelValueCombo *LabelValueCombo, value float64) {
+	g.change(labelValueCombo, value, set)
+}
+
+func (g *gauge) Inc(labelValueCombo *LabelValueCombo, value float64) {
+	g.change(labelValueCombo, value, add)
+}
+
+func (g *gauge) change(labelValueCombo *LabelValueCombo, value float64, operation string) {
```

Comment:
`change` is a strange name for this func. What about `updateSeries` instead, and `series` can have its own `update` method?